frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/GitFetchSupportImpl.java

```diff
@@ -0,0 +1,462 @@
+import io.vavr.control.Option;
+import io.vavr.control.Try;
+import kotlin.jvm.functions.Function1;
+import lombok.AllArgsConstructor;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import com.virtuslab.gitmachete.frontend.actions.toolbar.FetchAllRemotesAction;
+import com.virtuslab.logger.IPrefixedLambdaLogger;
+import com.virtuslab.logger.PrefixedLambdaLoggerFactory;
+
+/**
+ * @deprecated This implementation is a workaround to provide features missing in pre-2020.1 versions.
+ * The main benefit of it is specific branch update ({@link GitFetchSupportImpl#fetch(GitRepository, GitRemote, String)}).
+ * We need this class for {@link FetchAllRemotesAction} and {@link BasePullBranchAction}.
+ */
+@Deprecated
+public final class GitFetchSupportImpl implements GitFetchSupport {
+
+  private static final IPrefixedLambdaLogger LOG = PrefixedLambdaLoggerFactory.getLogger("frontendActions");
+
+  private static final int MAX_SSH_CONNECTIONS = 10;
+
+  private final Project project;
+  private final ProgressManager progressManager = ProgressManager.getInstance();
+
+  private AtomicInteger fetchRequestCounter = new AtomicInteger(0);
+
+  private GitFetchSupportImpl(Project project) {
+    this.project = project;
+  }
+
+  public static GitFetchSupportImpl fetchSupport(Project project) {
+    return new GitFetchSupportImpl(project);
+  }
+
+  public boolean isFetchRunning() {
+    return fetchRequestCounter.get() > 0;
+  }
+
+  @Override
+  public FetchResultImpl fetchDefaultRemote(Collection<GitRepository> repositories) {
+    var remotesToFetch = new ArrayList<RemoteRefCoordinates>();
+    for (var repository : repositories) {
+      var remote = getDefaultRemoteToFetch(repository);
+      if (remote != null) {
+        remotesToFetch.add(new RemoteRefCoordinates(repository, remote, /* refspec */ null));
+      } else {
+        LOG.info("No remote to fetch found in ${repository}");
+      }
+    }
+    return fetch(remotesToFetch);
+  }
+
+  @Override
+  public FetchResultImpl fetchAllRemotes(Collection<GitRepository> repositories) {
+    var remotesToFetch = new ArrayList<RemoteRefCoordinates>();
+    for (var repository : repositories) {
+      if (repository.getRemotes().isEmpty()) {
+        LOG.info("No remote to fetch found in ${repository}");
+      } else {
+        for (var remote : repository.getRemotes()) {
+          remotesToFetch.add(new RemoteRefCoordinates(repository, remote, /* refspec */ null));
+        }
+      }
+    }
+    return fetch(remotesToFetch);
+  }
+
+  public FetchResultImpl fetch(GitRepository repository, GitRemote remote, String refspec) {
+    return fetch(List.of(new RemoteRefCoordinates(repository, remote, refspec)));
+  }
+
+  @Override
+  public FetchResultImpl fetch(GitRepository repository, GitRemote remote) {
+    return fetch(List.of(new RemoteRefCoordinates(repository, remote, /* refspec */ null)));
+  }
+
+  public FetchResultImpl fetch(List<RemoteRefCoordinates> arguments) {
```

Comment:
This class already looks pretty race-condition prone, esp. if we're ever going to diverge from the original IJ implementation... pls replace Java collections with immutable Vavr counterparts wherever possible (preferably leave the Java collections only in the overridden method params and convert to immutable ones asap)

Reply:
done